pkg/core/middleware/authn.go

```diff
@@ -0,0 +1,131 @@
+package middleware
+
+import (
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/skygeario/skygear-server/pkg/core/auth"
+	"github.com/skygeario/skygear-server/pkg/core/auth/authinfo"
+	"github.com/skygeario/skygear-server/pkg/core/auth/session"
+	"github.com/skygeario/skygear-server/pkg/core/config"
+	"github.com/skygeario/skygear-server/pkg/core/db"
+	"github.com/skygeario/skygear-server/pkg/core/handler"
+	coreHttp "github.com/skygeario/skygear-server/pkg/core/http"
+	"github.com/skygeario/skygear-server/pkg/core/model"
+	"github.com/skygeario/skygear-server/pkg/core/skydb"
+	"github.com/skygeario/skygear-server/pkg/core/skyerr"
+)
+
+// AuthnMiddleware populate auth context information
+type AuthnMiddleware struct {
+	AuthContextSetter auth.ContextSetter `dependency:"AuthContextSetter"`
+	SessionProvider   session.Provider   `dependency:"SessionProvider"`
+	AuthInfoStore     authinfo.Store     `dependency:"AuthInfoStore"`
+	TxContext         db.TxContext       `dependency:"TxContext"`
+}
+
+// AuthnMiddlewareFactory creates AuthnMiddleware per request.
+type AuthnMiddlewareFactory struct{}
+
+// NewInjectableMiddleware implements InjectableMiddlewareFactory.
+func (f AuthnMiddlewareFactory) NewInjectableMiddleware() InjectableMiddleware {
+	return &AuthnMiddleware{}
+}
+
+// Handle implements InjectableMiddleware.
+func (m *AuthnMiddleware) Handle(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		var err error
+		defer func() {
+			if err == nil {
+				next.ServeHTTP(w, r)
+			} else {
+				// clear session cookie if error occurred
+				cookie := &http.Cookie{
+					Name:    coreHttp.CookieNameSession,
+					Path:    "/",
+					Expires: time.Unix(0, 0),
+				}
+				http.SetCookie(w, cookie)
+
+				skyErr := skyerr.NewNotAuthenticatedErr()
+				httpStatus := skyerr.ErrorDefaultStatusCode(skyErr)
+				response := handler.APIResponse{Err: skyErr}
+				encoder := json.NewEncoder(w)
+				w.Header().Set("Content-Type", "application/json")
+				w.WriteHeader(httpStatus)
+				encoder.Encode(response)
+			}
+		}()
+
+		tenantConfig := config.GetTenantConfig(r)
+
+		key := model.CheckAccessKey(tenantConfig, model.GetAPIKey(r))
+		m.AuthContextSetter.SetAccessKey(key)
+
+		accessToken, transport, err := model.GetAccessToken(r)
+		if err != nil {
+			// invalid session token -> must not proceed
+
+			if err == model.ErrTokenConflict {
+				// clear session cookie if session token is conflicted
+				cookie := &http.Cookie{
+					Name:    coreHttp.CookieNameSession,
+					Path:    "/",
+					Expires: time.Unix(0, 0),
+				}
+				http.SetCookie(w, cookie)
```

Comment:
This part seems redundant because in the `if err != nil` block, we never assign `err` to `nil` so the above defer block will clear the cookie for us.